### PR TITLE
NeroPlus optimizer

### PR DESCRIPTION
**Pull request by emailweixu**
This is an enhanced version of the Nero optimizer described in the following  paper:

Yang Liu et. al. "Learning by Turning: Neural Architecture Aware Optimisation" <https://arxiv.org/abs/2102.07227

The enhancement includes:

1. Option for ADAM like update (normalizing_grad_by_norm=False)
2. Upper bound contraint of weight norm (fixed_norm=False)
3. Weight decay

### CODE REVIEW

File: alf/optimizers/nero_plus.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+from typing import Tuple, Optional
+from .utils import get_opt_arg
+
+
+def _norm(x, max_norm):
+    if x.ndim > 1:
+        n = x.reshape(x.shape[0], -1).norm(dim=1)
+        return n.reshape(x.shape[0], *([1] * (x.ndim - 1)))
+    elif max_norm is not None:
+        return x.norm()
```
Comment (Haichao-Zhang) [inline alf/optimizers/nero_plus.py]: When ``max_norm`` is not None, based on its meaning, it seems that we should clamp the norm in this case?

File: alf/optimizers/nero_plus.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    `Yang Liu et. al. Learning by Turning: Neural Architecture Aware Optimisation
+    <https://arxiv.org/abs/2102.07227>`_
+
+    The essence of this optimizer is to keep the norm of each parameter vector
+    fixed and mean at zero during the optimization process. The parameter vector
+    is defined as the part of parameter responsible for one dimension of the
+    output. For example, ``FC(m, n)`` have two parameters, its weight of shape
+    [m, n] and its bias of shape [n]. Its weight have m parameter vectors. Each
+    of these m vectors is subject to the norm and mean constraint. For the bias,
+    one element is responsible for one output dimension. So it is not subject the
```
Comment (Haichao-Zhang) [inline alf/optimizers/nero_plus.py]: subject -> subject to

File: alf/optimizers/nero_plus.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            pmax_norm, pfixed_norm, pzero_mean = _get_opt_args(
+                p, max_norm, fixed_norm, zero_mean)
+            norm = _norm(p, pmax_norm)
+            if p.ndim > 1:
+                if pzero_mean:
+                    mean = p.mean(dim=tuple(range(1, p.ndim)))
+                    i = mean.abs().argmax()
+                    assert mean[i].abs() < 0.01, (
+                        "Unnormalized parameter: mean()=%s"
+                        "Model should be initialized using nero_initialize()" %
```
Comment (Haichao-Zhang) [inline alf/optimizers/nero_plus.py]: ``nero_initialize()``.  Seems need to update it with ``NeroPlus.initialize()``?

File: alf/optimizers/nero_plus.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+                        mean[i].item())
+                if pmax_norm is not None:
+                    diff = norm / pmax_norm - 1
+                    if pfixed_norm:
+                        diff.abs_()
+                    i = diff.argmax()
+                    assert diff[i] < 0.01, (
+                        "Unnormalized parameter: norm=%s. "
+                        "Model should be initialized using nero_initialize()" %
```
Comment (Haichao-Zhang) [inline alf/optimizers/nero_plus.py]: ``nero_initialize()`` ->  ``NeroPlus.initialize()``?

File: alf/optimizers/nero_plus.py (hunk 1 above)
Comment (runjerry) [reply alf/optimizers/nero_plus.py]: In current calls of _norm(), is there any case that ``max_norm`` is None?

File: alf/optimizers/nero_plus.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    The essence of this optimizer is to keep the norm of each parameter vector
+    fixed and mean at zero during the optimization process. The parameter vector
+    is defined as the part of parameter responsible for one dimension of the
+    output. For example, ``FC(m, n)`` have two parameters, its weight of shape
+    [m, n] and its bias of shape [n]. Its weight have m parameter vectors. Each
+    of these m vectors is subject to the norm and mean constraint. For the bias,
+    one element is responsible for one output dimension. So it is not subject the
+    norm and zero-mean constraint. Since the range of the output of a model should
+    not be constrained, you should set opt_config for the output layers as
+    `dict(fixed_norm=False, max_norm=math.inf)` or use a large finite ``max_norm``
```
Comment (runjerry) [inline alf/optimizers/nero_plus.py]: also zero_mean=False?

File: alf/optimizers/nero_plus.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                    assert mean[i].abs() < 0.01, (
+                        "Unnormalized parameter: mean()=%s"
+                        "Model should be initialized using nero_initialize()" %
+                        mean[i].item())
+                if pmax_norm is not None:
+                    diff = norm / pmax_norm - 1
+                    if pfixed_norm:
+                        diff.abs_()
+                    i = diff.argmax()
+                    assert diff[i] < 0.01, (
```
Comment (runjerry) [inline alf/optimizers/nero_plus.py:180]: Is this 0.01 also a hyper parameter in the Nero paper? Should we make it configurable?

File: alf/optimizers/nero_plus.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                else:
+                    exp_avg = grad
+                if normalizing_grad_by_norm:
+                    sq = _norm(grad, pmax_norm)**2
+                else:
+                    sq = grad**2
+                state['exp_avg_sq'].lerp_(sq, 1 - beta2)
+                denom = state['exp_avg_sq'].sqrt().add_(eps)
+                step_size = lr_scale * lr * math.sqrt(
+                    bias_correction2) / bias_correction1
```
Comment (runjerry) [inline alf/optimizers/nero_plus.py:256]: Could you add a brief comment regarding this ``step_size`` computation?

File: alf/optimizers/nero_plus_test.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            alf.layers.Conv2D(32, 32, 3, strides=2, padding=1),
+            alf.layers.Reshape(-1),
+            alf.layers.FC(
+                4 * 4 * 32,
+                num_classes,
+                weight_opt_args=dict(
+                    fixed_norm=False, weight_decay=1e-3,
+                    max_norm=float('inf'))))
+        NeroPlus.initialize(model)
+        opt = NeroPlus(lr=0.01)
```
Comment (runjerry) [inline alf/optimizers/nero_plus_test.py]: Do you want to also include test cases beyond the default setting, say original Nero and without running average?

File: alf/optimizers/nero_plus.py (hunk 5 above)
Comment (emailweixu) [reply alf/optimizers/nero_plus.py]: Good catch

File: alf/optimizers/nero_plus.py (hunk 6 above)
Comment (emailweixu) [reply alf/optimizers/nero_plus.py:180]: This check is used to ensure NeroPlus.initialize() is called to normalize the weights before adding them to the normalizer. A slight error is expected due to limited numerical precision. And all the weights will be normalized during the subsequent optimization steps. So it does not matter.

File: alf/optimizers/nero_plus.py (hunk 7 above)
Comment (emailweixu) [reply alf/optimizers/nero_plus.py:256]: Added.

File: alf/optimizers/nero_plus.py (hunk 1 above)
Comment (emailweixu) [reply alf/optimizers/nero_plus.py]: You are right. Removed max_norm from arguments.

File: alf/optimizers/nero_plus_test.py (hunk 8 above)
Comment (emailweixu) [reply alf/optimizers/nero_plus_test.py]: added

**Review by Haichao-Zhang — APPROVED** [review]

File: alf/optimizers/nero_plus.py (hunk 6 above)
Comment (runjerry) [reply alf/optimizers/nero_plus.py:180]: Got it, thanks for the explaination.

**Review by runjerry — APPROVED** [review]

**Pull request merged**